File: pluk/src/ssh/agent.ts
import { createConnection } from "net";
import { existsSync } from "fs";
import { homedir } from "os";
import { parseSSHConfig } from "./config.js";

// SSH agent reachability. Auth through an agent can only work — and an approval
// prompt can only appear — if requests actually reach a live agent. A
// GUI-launched app inherits whatever SSH_AUTH_SOCK launchd set, which is often
// an empty or stale agent, while the user's 1Password agent listens on its own
// well-known socket. So before claiming anything about approvals, probe the
// candidates with a real agent request (SSH_AGENTC_REQUEST_IDENTITIES) and pick
// the first one that can plausibly sign. The probe doubles as the wake-up call:
// a locked 1Password pops its unlock window on exactly this kind of request.

export const SSH_AGENT_UNREACHABLE_CODE = "SSH_AGENT_UNREACHABLE";

const PROBE_TIMEOUT_MS = 2_000;

const SSH_AGENTC_REQUEST_IDENTITIES = 11;
const SSH_AGENT_IDENTITIES_ANSWER = 12;

export type AgentProbe =
  | { state: "keys"; keys: number } // answered the identity list, has keys
  | { state: "empty" } // answered with zero keys — reachable but can't sign
  | { state: "mute" } // connected but no answer — e.g. a locked agent holding a prompt
  | { state: "dead"; error: string };

export function probeAgentSocket(path: string, timeoutMs = PROBE_TIMEOUT_MS): Promise<AgentProbe> {
  return new Promise((resolve) => {
    const sock = createConnection(path);
    let connected = false;
    let buf = Buffer.alloc(0);
    let done = false;
    const finish = (result: AgentProbe) => {
      if (done) return;
      done = true;
      clearTimeout(timer);
      sock.destroy();
      resolve(result);
    };
    const timer = setTimeout(
      () => finish(connected ? { state: "mute" } : { state: "dead", error: "connect timed out" }),
      timeoutMs
    );
    sock.on("connect", () => {
      connected = true;
      sock.write(Buffer.from([0, 0, 0, 1, SSH_AGENTC_REQUEST_IDENTITIES]));
    });
    sock.on("data", (chunk: Buffer) => {
      buf = Buffer.concat([buf, chunk]);
      if (buf.length < 5) return;
      if (buf[4] !== SSH_AGENT_IDENTITIES_ANSWER) return finish({ state: "mute" });
      if (buf.length < 9) return;
      const keys = buf.readUInt32BE(5);
      finish(keys > 0 ? { state: "keys", keys } : { state: "empty" });
    });
    sock.on("error", (e) => finish({ state: "dead", error: e.message }));
    sock.on("close", () => finish({ state: "dead", error: "agent closed the connection" }));
  });
}

function wellKnownAgentSockets(): string[] {
  return [
    `${homedir()}/Library/Group Containers/2BUA8C4S2C.com.1password/t/agent.sock`,
    `${homedir()}/.1password/agent.sock`,
  ].filter(existsSync);
}

export function agentSocketCandidates(host: string): string[] {
  const fromConfig = parseSSHConfig(host).identityAgent;
  const all = [fromConfig, ...wellKnownAgentSockets(), process.env.SSH_AUTH_SOCK]
    .filter((p): p is string => Boolean(p));
  return [...new Set(all)];
}

export interface LiveAgent {
  socket: string;
  probe: AgentProbe;
}

// Pick the agent socket most likely to complete a signature: one that lists
// keys wins immediately. Otherwise a mute one (a locked 1Password — the probe
// just asked it to unlock) is worth waiting on, since signing can still wake
// it into prompting. An agent that answered with zero keys can never sign, so
// it is never picked over nothing — a dead one can neither sign nor prompt.
export function pickLiveAgent(probed: LiveAgent[]): LiveAgent | undefined {
  return probed.find((p) => p.probe.state === "mute");
}

export async function resolveLiveAgent(host: string): Promise<LiveAgent | undefined> {
  const probed: LiveAgent[] = [];
  for (const socket of agentSocketCandidates(host)) {
    const probe = await probeAgentSocket(socket);
    console.log(
      `[pluk] SSH agent probe: ${socket} -> ${probe.state}${probe.state === "dead" ? ` (${probe.error})` : ""}`
    );
    if (probe.state === "keys") return { socket, probe };
    if (probe.state !== "dead") probed.push({ socket, probe });
  }
  return pickLiveAgent(probed);
}

export function agentUnreachableError(): Error {
  const err = new Error(
    "Can't reach your SSH key agent — no agent socket answered, so no approval prompt can appear. " +
    "Open and unlock 1Password (with its SSH agent enabled), or load the key into ssh-agent, then retry."
  );
  (err as Error & { code: string }).code = SSH_AGENT_UNREACHABLE_CODE;
  return err;
}

File: crates/pluk-ssh/src/agent.rs

pub const SSH_AGENT_UNREACHABLE_CODE: &str = "SSH_AGENT_UNREACHABLE";

const PROBE_TIMEOUT_MS: u64 = 2_000;
const SSH_AGENTC_REQUEST_IDENTITIES: u8 = 11;
const SSH_AGENT_IDENTITIES_ANSWER: u8 = 12;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum AgentProbe {
    Keys { keys: u32 },
    Empty,
    Mute,
    Dead { error: String },
}

impl AgentProbe {
    pub fn state_str(&self) -> &'static str {
        match self {
            Self::Keys { .. } => "keys",
            Self::Empty => "empty",
            Self::Mute => "mute",
            Self::Dead { .. } => "dead",
        }
    }
}

#[derive(Debug, Clone)]
pub struct LiveAgent {
    pub socket: String,
    pub probe: AgentProbe,
}

/// Probe an agent socket by actually connecting and sending
/// `SSH_AGENTC_REQUEST_IDENTITIES`. The probe itself wakes a locked 1Password,
/// so it must connect rather than just stat the path.
pub async fn probe_agent_socket(path: &str, timeout_ms: u64) -> AgentProbe {
    let timeout = std::time::Duration::from_millis(timeout_ms);
    tokio::time::timeout(timeout, probe_inner(path))
        .await
        .unwrap_or(AgentProbe::Dead {
            error: "connect timed out".into(),
        })
}

async fn probe_inner(path: &str) -> AgentProbe {
    use tokio::io::{AsyncReadExt, AsyncWriteExt};

    let mut stream = match tokio::net::UnixStream::connect(path).await {
        Ok(s) => s,
        Err(e) => {
            return AgentProbe::Dead {
                error: e.to_string(),
            }
        }
    };

    // Send SSH_AGENTC_REQUEST_IDENTITIES: 4-byte len + 1 byte type
    let req = [0u8, 0, 0, 1, SSH_AGENTC_REQUEST_IDENTITIES];
    if let Err(e) = stream.write_all(&req).await {
        return AgentProbe::Dead {
            error: e.to_string(),
        };
    }

    // Wait for response with a short read timeout
    let mut buf = vec![0u8; 8192];
    let read_fut = stream.read(&mut buf);
    let n = match tokio::time::timeout(std::time::Duration::from_millis(PROBE_TIMEOUT_MS), read_fut).await {
        Ok(Ok(0)) => {
            return AgentProbe::Dead {
                error: "agent closed the connection".into(),
            }
        }
        Ok(Ok(n)) => n,
        Ok(Err(e)) => {
            return AgentProbe::Dead {
                error: e.to_string(),
            }
        }
        Err(_) => return AgentProbe::Mute,
    };

    if n < 5 {
        return AgentProbe::Mute;
    }
    if buf[4] != SSH_AGENT_IDENTITIES_ANSWER {
        return AgentProbe::Mute;
    }
    if n < 9 {
        return AgentProbe::Mute;
    }
    let keys = u32::from_be_bytes([buf[5], buf[6], buf[7], buf[8]]);
    if keys > 0 {
        AgentProbe::Keys { keys }
    } else {
        AgentProbe::Empty
    }
}

fn well_known_agent_sockets() -> Vec<String> {
    let home = match pluk_core::platform::home_dir() {
        Some(h) => h,
        None => return vec![],
    };
    let candidates = [
        home.join("Library/Group Containers/2BUA8C4S2C.com.1password/t/agent.sock"),
        home.join(".1password/agent.sock"),
    ];
    candidates
        .into_iter()
        .filter(|p| p.exists())
        .map(|p| p.to_string_lossy().to_string())
        .collect()
}

pub fn agent_socket_candidates(host: &str) -> Vec<String> {
    let from_config = crate::config::parse_ssh_config(host)
        .identity_agent
        .filter(|s| !s.is_empty());
    let mut all: Vec<String> = Vec::new();
    if let Some(s) = from_config {
        all.push(s);
    }
    all.extend(well_known_agent_sockets());
    if let Ok(sock) = std::env::var("SSH_AUTH_SOCK")
        && !sock.is_empty() {
            all.push(sock);
        }
    // Deduplicate preserving order
    let mut seen = std::collections::HashSet::new();
    all.into_iter().filter(|s| seen.insert(s.clone())).collect()
}

/// Prefer an agent that answered with keys; otherwise a mute one (locked
/// 1Password — the probe just asked it to unlock). Empty/dead are not picked.
pub fn pick_live_agent(probed: &[LiveAgent]) -> Option<LiveAgent> {
    // First try to find a mute one (locked 1Password — worth waiting on)
    probed.iter().find(|p| p.probe == AgentProbe::Mute).cloned()
}

pub async fn resolve_live_agent(host: &str) -> Option<LiveAgent> {
    let mut probed: Vec<LiveAgent> = Vec::new();
    for socket in agent_socket_candidates(host) {
        let probe = probe_agent_socket(&socket, PROBE_TIMEOUT_MS).await;
        eprintln!(
            "[pluk] SSH agent probe: {} -> {}{}",
            socket,
            probe.state_str(),
            match &probe {
                AgentProbe::Dead { error } => format!(" ({error})"),
                _ => String::new(),
            }
        );
        match &probe {
            AgentProbe::Keys { .. } => return Some(LiveAgent { socket, probe }),
            AgentProbe::Dead { .. } => {}
            _ => probed.push(LiveAgent {
                socket: socket.clone(),
                probe,
            }),
        }
    }
    pick_live_agent(&probed)
}

#[derive(Debug, thiserror::Error)]
#[error("{}", message)]
pub struct AgentUnreachableError {
    pub message: String,
    pub code: &'static str,
}

pub fn agent_unreachable_error() -> AgentUnreachableError {
    AgentUnreachableError {
        message: "Can't reach your SSH key agent — no agent socket answered, so no approval prompt can appear. Open and unlock 1Password (with its SSH agent enabled), or load the key into ssh-agent, then retry.".into(),
        code: SSH_AGENT_UNREACHABLE_CODE,
    }
}

impl AgentUnreachableError {
    pub fn to_std_error(&self) -> std::io::Error {
        let e = std::io::Error::new(std::io::ErrorKind::NotFound, self.message.clone());
        // Attach code via display; callers check via is_ssh_auth_error
        e
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pick_prefers_mute_over_empty() {
        let agents = vec![
            LiveAgent {
                socket: "/tmp/a.sock".into(),
                probe: AgentProbe::Empty,
            },
            LiveAgent {
                socket: "/tmp/b.sock".into(),
                probe: AgentProbe::Mute,
            },
        ];
        assert_eq!(pick_live_agent(&agents).unwrap().socket, "/tmp/b.sock");
    }

    #[test]
    fn pick_returns_none_when_all_dead_or_empty() {
        let agents = vec![LiveAgent {
            socket: "/tmp/a.sock".into(),
            probe: AgentProbe::Empty,
        }];
        assert!(pick_live_agent(&agents).is_none());
        assert!(pick_live_agent(&[]).is_none());
    }

    #[test]
    fn agent_unreachable_has_code() {
        let e = agent_unreachable_error();
        assert_eq!(e.code, SSH_AGENT_UNREACHABLE_CODE);
        assert!(e.message.contains("1Password"));
    }

    #[tokio::test]
    async fn probe_nonexistent_is_dead() {
        let probe = probe_agent_socket("/tmp/pluk-test-nonexistent-xyz.sock", 200).await;
        assert!(matches!(probe, AgentProbe::Dead { .. }));
    }
}
